RU CSC segment builder parameters initialization fix

nvoytish opened the pull request:
Raise in the issue:
https://github.com/cms-sw/cmssw/issues/18605#issuecomment-302016415

the parameters of the RU CSC segment builder were not initialized in the right way, causing difference in reconstruction while running in multithreading mode.
The parameters were switched to local and are initialized properly.

nvoytish committed:
Update CSCSegmentAlgorithmRU_cfi.py

unused parameter removed

nvoytish committed:
Update CSCSegAlgoRU.cc

parameters initialized in a proper way

nvoytish committed:
Parameter initialization fixed

Parameter initialization fixed
Unused variables deleted

nvoytish committed:
print removed

cmsbuild commented:
A new Pull Request was created by  @nvoytish  for master.

It involves the following packages:

RecoLocalMuon/CSCSegment


@perrotta, @cmsbuild, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@ptcox, @bellan, @abbiendi, @jhgoh this is something you requested to watch as well.
@davidlange6 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 1c8e9a23c8d35ef4bfab94b7de799a17c3594d8d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19081/20311/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7d62dba5b066e250c737d8b05bd8ca5992fbc820
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19081/20311/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19081/20311/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19081/20311/summary.html

Comparison Summary:
 - You potentially added 1318 lines to the logs
 - Reco comparison results: 3423 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1837639
 - DQMHistoTests: Total failures: 34098
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1803368
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

perrotta commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -315,6 +313,7 @@ std::vector<CSCSegment> CSCSegAlgoRU::buildSegments(const ChamberHitContainer& u
       ++it;//go to the next seg
     }
   }//while
+  std::cout<<" segs built "<<std::endl;
```
please remove this cout

perrotta commented:
@nvoytish : could you please point me to the actual fixes implemented with this PR?
 I see some renamed variable, and a few removed ones (quite likely not needed), but I can't understand where the actual fix is...

nvoytish committed:
unneded cout removed

cmsbuild commented:
Pull request #19081 was updated. @perrotta, @cmsbuild, @slava77, @davidlange6 can you please check and sign again.

nvoytish commented:
The variables used inside the RU class had same names as the parameters from the python file and sometimes were not properly reset.

perrotta commented:
Can you please show me one example of one of those local variables with the same name as the parameters? I can't find it... (my bad, of course)

In the meanwhile, you could remove the underscore at the end of l_chi2_str_ and l_chi2Norm_2D_ , which is not needed any more to distinguish chi2_str_  from chi2_str

nvoytish commented:
all parameters from RecoLocalMuon/CSCSegment/python/CSCSegmentAlgorithmRU_cfi.py are handed to the class starting at:
https://github.com/cms-sw/cmssw/blob/master/RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc#L20
all of them had same names as the variables. I simply added "l_" to their name so they differ.

Thank you for reviewing!

nvoytish commented:
> you could remove the underscore at the end of l_chi2_str_ and l_chi2Norm_2D_ , which is not needed any more to distinguish chi2_str_ from chi2_str

I didn't quite get what you meant. Could you explain.

perrotta commented:
nvoytish <notifications@github.com> ha scritto:

> all parameters from  
> RecoLocalMuon/CSCSegment/python/CSCSegmentAlgorithmRU_cfi.py are  
> handed to the class starting at:
> https://github.com/cms-sw/cmssw/blob/master/RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc#L20
> all of them had same names as the variables. I simply added "l_" to  
> their name so they differ.
>

I don't see where is the problem if a local variable in a class has  
the same name as one parameter in a python config...


>> you could remove the underscore at the end of l_chi2_str_ and  
>> l_chi2Norm_2D_ , which is not needed any more to distinguish  
>> chi2_str_ from chi2_str
>
> I didn't quite get what you meant. Could you explain.
>

This is really a minor point, but it would be nice to keep the same  
style for the names of the variables. If you already identify the  
parameters taken from the cfg with an inital "l_", there is no need to  
identify just two of them with a final "_". This was sort of needed  
before because you had both "chi2_str_" and "chi2_str"; but now you  
have "l_chi2_str_" and "chi2_str", which can more conveniently become  
"l_chi2_str" and "chi2_str"

nvoytish commented:
I got your idea. I'll change this. 

Could everybody keep this PR on hold. 
I changed actually two things:
 a) rennemed the variables;
 b) switch back from using unique_ptr to the way CSCSegFit is used in other CSC segment builders, for example here https://github.com/cms-sw/cmssw/blob/master/RecoLocalMuon/CSCSegment/src/CSCSegAlgoSK.cc#L394

these two changes were giving me identical result while running multithreading jobs.
As discussed in https://github.com/cms-sw/cmssw/issues/18605#issuecomment-302016415

I got the idea that I am using unique_ptr in the rignt way, so I removed the changes from the point b).
But it looks like I am not using unique_ptr in the rignt way, because I get different results with the code from the PR .

Could anybody look at the usage of the CSCSegFit in the commited code, while I am also searching for the problem?

Thank you!

ptcox commented:
Nikolay, I am confused. 

I thought the issue was arising from the change made to parameters set from config, which did not always get reset between events to the exact same starting values?

Are you saying
1) the identical results you showed in last Friday's Reco Meeting had both the parameters reset correctly, AND unique_ptr removed?
2) If you fix the parameters but use unique_ptr you still see problems with multithreading?

Assuming the use of CSCSegFit* __as a member variable__ of the algorithm is correct in another algorithm is not robust because the old algorithm was never tested with multithreading! Notice that the standard 'old' algorithm was ST, and does NOT have CSCSegFit* as a member.

nvoytish commented:
The answer to both your questions is "yes". Before submitting the PR I asked Slava about the unique_ptr usage and we agreed that I use it in right way.

I just double checked - there is no difference in the results if I drop unque_ptr and use it in the "old fasion" way.

ptcox commented:
That's frustrating. It would seem to me that the parameters issue MUST have given rise to some problems. I'm not an expert on unique_ptr so I can't help with that. I notice the code still does some setting of sfit_ to 0 (which would be done for a bare pointer only after a delete.) Not clear to me if that makes sense for a unique_ptr?

ptcox commented:
Not sure about the code in compareProtoSegment and increaseProtoSegment, like
    `proto_segment = oldproto;`
    `sfit_ = std::move(oldfit); // reset to the original input fit`
Doesn't one have to delete the old object first? i.e. `sfit_.reset();` 
Or is it smart enough to do that automatically?
I'd remove the` sfit_= 0;` on lines 98 and 235 too, and assume unique_ptr is keeping control properly.
Does the logic of lines 161, 
 `if ( sfit_ ) tryAddingHitsToSegment(rechits, used, layerIndex, i1, i2);`
and 478
` if ( !sfit_ ) return 0.;`
work with unique_ptr?
Let's hope some expert can cross check.

nvoytish commented:
As far as I know it takes care automatically about deleting itself. 
I'll try what you suggested.

nvoytish commented:
I gave a try to your suggestions - no luck.

perrotta commented:
I think this PR got stuck. As such it does absolutely nothing but renaming a few local variables.

I have tried to look at the code of the files touched here, but I couldn't find anything obviously related to some possible non-reproducibility in multithreaded runs (there is the initialization of a unique_ptr to "0" which is not clear to me what it actually does, as also noticed by Tim: it is certainly more apropriate reset()'ting it).

Do you plan to update it any soon? Did you explore all suggestions raised in issue #18605? (I will try to go through that git issue once again, possibly tomorrow: but it will certainly be more effective if you, the experts, try to go through it and look for the logic bug in the code)

ptcox commented:
We're working on it. No logic error has been found (yet). Unfortunately it seems that using a bare pointer for CSCSegFit works while using a unique_ptr doesn't. Which makes us feel there is some issue we're missing in the pointer handling.I've asked Nikolay to fix the config params and avoid manipulating them because that is obviously something to simplify. I've also asked him to provide corresponding plots to those he showed at last Friday's Reco Meeting where no problem is visible, but where the problem can actually be seen. We still don't know for sure if the problem arises from an initialization issue or from  pointer mis-use.


On June 8, 2017 at 16:30:30, perrotta (notifications@github.com) wrote:

I think this PR got stuck. As such it does absolutely nothing but renaming a few local variables.

I have tried to look at the code of the files touched here, but I couldn't find anything obviously related to some possible non-reproducibility in multithreaded runs (there is the initialization of a unique_ptr to "0" which is not clear to me what it actually does, as also noticed by Tim: it is certainly more apropriate reset()'ting it).

Do you plan to update it any soon? Did you explore all suggestions raised in issue #18605? (I will try to go through that git issue once again, possibly tomorrow: but it will certainly be more effective if you, the experts, try to go through it and look for the logic bug in the code)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

perrotta commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -118,16 +116,16 @@ std::vector<CSCSegment> CSCSegAlgoRU::buildSegments(const ChamberHitContainer& u
     //change the tresholds if it's time to look for displaced mu segments
-    if(doCollisions && search_disp && int(rechits.size()-used_rh)>2){//check if there are enough recHits left to build a segment from displaced vertices
-      doCollisions = false;
+    if(l_doCollisions && search_disp && int(rechits.size()-used_rh)>2){//check if there are enough recHits left to build a segment from displaced vertices
```
I can't imagine it can be responsible of any non-reproducibility.

Anyhow this line is potentially wrong: "rechits.size()" is unsigned, therefore the difference can be unpredictable if used_rh>rechits.size() (can it happen?). 

It should be better rewritten as either
- int(rechits.size())-used_rh>2
- int(rechits.size())>used_rh+2

ptcox replied:
Andrea, good eyes. I think this is strictly a bug as you say. But used_rh should be converted to unsigned before the subtraction and I think that will be fine: rechits.size() is all the rechits available to process so it is always larger than the number 'used' in a segment.

I'm still of the opinion that non-reproducibility is most likely from the initialization manipulations, and Nikolay is working on clearing that up.

slava77 replied:
by construction, used_rh is a count of hits passing some selections
see a few lines above (slightly re-parsed):
```C++
    int used_rh = 0;
    for (ChamberHitContainerCIt i1 = rechits.begin(); i1 != rechits.end(); ++i1) {
      if(used[i1-ib])used_rh++;
    }
```
so, it can not be larger than rechits.size().

perrotta replied:
Sure.
But nothing prevent us writing that line of code in correct C++

slava77 replied:
I'm missing the point though: what is wrong with the syntax or the expression?
if size() is smaller than used_rh, it will still produce a correct signed answer, at least with the normal range of values (both sufficiently below 2^31-1).

Please note that the type of size() is 64-bit unsigned.
So, in a more abstract case, where size() is larger than 2^31-1, the proposed change will more likely produce an incorrect result

slava77 commented:
@nvoytish @ptcox 
please clarify on the progress in this PR.
Thank you.

nvoytish commented:
We are trying to make the RU algo to use CSCSegFit in the right way. 
At the moment it doesn't. Unfortunately this week I am at a conference as its organizer and the Internet connection is bad here (working on lxplus machines you lose the connection every 10-15 minutes). Tim had a suggestion that I am trying to work on at the moment.

Best regards,
Nikolay Voytishin

________________________________________
From: Slava Krutelyov [notifications@github.com]
Sent: 14 June 2017 14:09
To: cms-sw/cmssw
Cc: Nikolay Voytishin; Mention
Subject: Re: [cms-sw/cmssw] RU CSC segment builder parameters initialization fix (#19081)

@nvoytish<https://github.com/nvoytish> @ptcox<https://github.com/ptcox>
please clarify on the progress in this PR.
Thank you.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/19081#issuecomment-308411613>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AFgBEgjaCXe_BEI_jpXt6ohrL8KFQFjlks5sD810gaJpZM4NuVht>.

ptcox commented:
And I will review the code again when I get some time. Today was a Muon DOC3 tutorial and we have a Muon Upgrade Workshop from today until Saturday.Then CMS Week with the usual pressure to supply slides for endless presentations. We're stretched too thin.

________________________________
From: nvoytish [notifications@github.com]
Sent: 14 June 2017 20:27
To: cms-sw/cmssw
Cc: Tim Cox; Mention
Subject: Re: [cms-sw/cmssw] RU CSC segment builder parameters initialization fix (#19081)

We are trying to make the RU algo to use CSCSegFit in the right way.
At the moment it doesn't. Unfortunately this week I am at a conference as its organizer and the Internet connection is bad here (working on lxplus machines you lose the connection every 10-15 minutes). Tim had a suggestion that I am trying to work on at the moment.

Best regards,
Nikolay Voytishin

________________________________________
From: Slava Krutelyov [notifications@github.com]
Sent: 14 June 2017 14:09
To: cms-sw/cmssw
Cc: Nikolay Voytishin; Mention
Subject: Re: [cms-sw/cmssw] RU CSC segment builder parameters initialization fix (#19081)

@nvoytish<https://github.com/nvoytish> @ptcox<https://github.com/ptcox>
please clarify on the progress in this PR.
Thank you.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/19081#issuecomment-308411613>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AFgBEgjaCXe_BEI_jpXt6ohrL8KFQFjlks5sD810gaJpZM4NuVht>.


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/19081#issuecomment-308518340>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AE2FnqkFgPfWXK8W8yhYoXdvf4J-3YU3ks5sECYtgaJpZM4NuVht>.

slava77 commented:
On 6/14/17 11:45 AM, ptcox wrote:
> And I will review the code again when I get some time. Today was a Muon
> DOC3 tutorial and we have a Muon Upgrade Workshop from today until
> Saturday.Then CMS Week with the usual pressure to supply slides for
> endless presentations. We're stretched too thin.

Hi Tim,

If the issue is not resolved soon, we will have to go back to the old
algo (which was known to be reproducible) for the physics production
release.
The time scale for that is about a week, maybe less.



> 
> ________________________________
> From: nvoytish [notifications@github.com]
> Sent: 14 June 2017 20:27
> To: cms-sw/cmssw
> Cc: Tim Cox; Mention
> Subject: Re: [cms-sw/cmssw] RU CSC segment builder parameters
> initialization fix (#19081)
> 
> We are trying to make the RU algo to use CSCSegFit in the right way.
> At the moment it doesn't. Unfortunately this week I am at a conference
> as its organizer and the Internet connection is bad here (working on
> lxplus machines you lose the connection every 10-15 minutes). Tim had a
> suggestion that I am trying to work on at the moment.
> 
> Best regards,
> Nikolay Voytishin
> 
> ________________________________________
> From: Slava Krutelyov [notifications@github.com]
> Sent: 14 June 2017 14:09
> To: cms-sw/cmssw
> Cc: Nikolay Voytishin; Mention
> Subject: Re: [cms-sw/cmssw] RU CSC segment builder parameters
> initialization fix (#19081)
> 
> @nvoytish<https://github.com/nvoytish> @ptcox<https://github.com/ptcox>
> please clarify on the progress in this PR.
> Thank you.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on
> GitHub<https://github.com/cms-sw/cmssw/pull/19081#issuecomment-308411613>,
> or mute the
> thread<https://github.com/notifications/unsubscribe-auth/AFgBEgjaCXe_BEI_jpXt6ohrL8KFQFjlks5sD810gaJpZM4NuVht>.
> 
>

perrotta commented:
Hi all: any update on this PR?

ptcox commented:
I heard from Nikolay today that he couldn't finish the reworking of initialization and logic this week because of institute workload, and he will be away next week because of a family issue. I see no rapid solution (we don't even know if the above work WILL cure the issue) so best revert to CSCSegAlgoST if it is a show-stopper.


On June 22, 2017 at 15:57:37 , perrotta (notifications@github.com) wrote:

Hi all: any update on this PR?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

slava77 commented:
-1
superseded by #19421 
IIUC, there is still a follow up expected with improvements to the quality of the code and perhaps a better understanding of the reasons that lead to non-reproducibility.
